synapse/storage/data_stores/main/registration.py

```diff
@@ -158,21 +158,44 @@ def set_account_validity_for_user_txn(txn):
         )
 
     @defer.inlineCallbacks
-    def get_expired_users(self):
-        """Get IDs of all expired users
+    def get_expired_users(self, active_only: bool = False):
+        """Get UserIDs of all expired users
+
+        Args:
+            active_only: Whether to return only those users that are marked as active
+                as part of their profile (note that this flag is separate from entire
+                account deactivation)
 
         Returns:
-            Deferred[list[str]]: List of expired user IDs
+            Deferred[List[UserID]]: List of expired user IDs
         """
 
         def get_expired_users_txn(txn, now_ms):
+            sql_args = (now_ms,)
+
+            # Determine whether we should constrain our query based on whether
+            # the user is active
+            join_sql = ""
+            if active_only:
+                join_sql = """
+                LEFT JOIN profiles as p
+                ON av.user_id = p.user_id
+                AND p.active = ?
```

Comment:
This will just return `NULL` for every column from `profiles` that doesn't match, instead of limiting to the value of `p.active`. You probably want to move the check on `p.active` to the `WHERE` clause.

Reply:
Ah, great spot thanks!

Reply:
Hm, oddly in my tests I don't get `null` back from this. In fact moving it to outside of the join ends up with no rows ever being returned :thinking:

Reply:
So it turns out that there was an issue due to comparing `profiles.user_id` (which is a localpart...) and `account_validity.user_id` which is a fully qualified user ID. So profiles would never be returned.

`AccountValidityUserDirectoryTestCase.test_expired_user_in_directory` was a good case to test this, but it only tested that we weren't replicating a profile at the end of everything. It should've also tested that we did replicate a profile before we expired the user.

So I've added that now.